rholang/src/main/k/rholang.k

```diff
@@ -577,40 +609,64 @@ rule <k> #(MATCHFAIL) ~> #(MATCHPAUSE) ~> match P:HigherProcOrChan { L:HigherMat
-// THE DEFINITION OF THE "match" FUNCTION (NOT the match process)
+// THE DEFINITION OF THE GENERAL "match" FUNCTION (NOT the match process)
+// This is the main engine for pattern matching. We will need this same algorithm
+// in slightly nuanced forms throughout the framework, so we attach the identifying
+// string to each instance of the match function.
 // ******* ******* ******* ******* ******* ******* ******* ******* ******* ******* ******* ******* ******* ******* //
 
 // Take the tuples one at a time
-rule <thread> ... <k> ... listmatch( P:HigherProc, Q:HigherProcs ; D:Chan, E:Chans )
-                      => listmatch( Q ; E ) ~> match( @P ; D )  ... </k> ... </thread>
+rule <thread> ... <k> ... listmatch[ S:String ]( P:HigherProc, Q:HigherProcs ; D:Chan, E:Chans )
+                      => listmatch[ S ]( Q ; E ) ~> match[S]( @P ; D )  ... </k> ... </thread>
 
 // When they're both done, get rid of the listmatch bit and get going on matching
-rule <thread> ... <k> ... listmatch( .HigherProcs ; .Chans )
+rule <thread> ... <k> ... listmatch[ S:String ]( .HigherProcs ; .Chans )
                       => . ... </k> ... </thread>
 
 // They fail if the tuples are not the same length
-rule <thread> ... <k> ... listmatch( .HigherProcs ; D:Chan, E:Chans )
-                      => #(MATCHFAIL) ... </k> ... </thread>
-rule <thread> ... <k> ... listmatch( P:HigherProc, Q:HigherProcs ; .Chans )
-                      => #(MATCHFAIL) ... </k> ... </thread>
-
-// (a) Give the rules for the match action which receives two arguments
-rule <thread> ... <k> ... match( P:HigherProcOrChan ; D:HigherProcOrChanNoVars )
-              => matchstdform( intostdmatchform(P) ; intostdmatchform(D)) ... </k> ... </thread>
-
-
-// If a variable matches with something, check if that thing has free variables
-// If it does, then rewrite as #(MATCHFAIL). If it does not, then substitute S for X
-// in the original statement. Checking for free variables will come soon.
-// For now, we just substitute. The code we would use for matching is below, once
-// semantics for the "freevars( )" function is done.
-rule <thread> ... <k> ... match( P:HigherProcOrChan ; X:Id )
-              => . ... </k>
-              <subst> Q => Q[ P / X] </subst> ... </thread>
-
-//rule <thread> ... <k> ... match(S:Stmt, X:Id) => freevars(S,X) ... </k> ... </thread>
-
-
+rule <thread> ... <k> ... listmatch[ S:String ]( .HigherProcs ; D:Chan, E:Chans )
+                      => #(MATCHFAIL)[S] ... </k> ... </thread>
+rule <thread> ... <k> ... listmatch[ S:String ]( P:HigherProc, Q:HigherProcs ; .Chans )
+                      => #(MATCHFAIL)[S] ... </k> ... </thread>
+
+// We also split things up if there are logical connectives: "and's" or "or's"
+// (i) The "and's" are straightforward, since we need to match with all of them
+rule <thread> ... <k> ... match[S:String]( P:HigherProcOrChan ; C:Chan /\ D:Chan )
+               => match[S](P ; C) ~> match[S]( P ; D ) ... </k> ... </thread>
```

Comment:
Does matching "short circuit" here?  That is, if `match[S](P ; C)` fails will it still try to `match[S]( P ; D )`?

Reply:
No, it won't. It will delete the second match checker and say that the patterns did not match.

Reply:
It rewrites to `#(MATCHFAIL)[ String ]` which triggers that.